---
title: "intro-distQTL"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{intro-distQTL}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup}
library(distQTL)
```

# Introduction

This vignette demonstrates **dist**ribution-based **q**uantitative **t**rait
**l**ocus mapping as implemented by the `R` package `distQTL`.  distQTL-mapping
is performed by way of Fréchet regression for univariate distribution response
objects, and partial $F$ tests designed for those settings.  The `R` package
`distQTL` utilizes the `R` package `fastfrechet` as a Fréchet regression
workhorse, and implements its own partial $F$ test in the function
`Wasserstein_F`. The `R` package `fastfrechet` should be automatically
installed as part of `distQTL`; its development version is currently available
for download from GitHub
[here](<https://github.com/alexandercoulter/fastfrechet>).

`disTQTL` implements distQTL-mapping through the eponymous function `distQTL`
(i.e. `distQTL::distQTL(...)`). The function expects a collection of
`data.table` structure inputs, which contain

1. single nucleotide polymorphism (SNP) genotype information for a set of
donors;
2. single cell RNA-sequence (scRNA-seq) expression data for a set of genes;
3. fixed donor specific covariate vectors;
4. cell-specific information, including donor ID label and cell type label; and
5. gene/SNP information, including names/IDs, chromosome locations, and genotype
start loci.

This vignette illustrates how these input objects should be structured. Other
inputs, such as donor/gene/SNP filtering specifications and desired cell type
groupings, are also explained and illustrated.

The main result of `distQTL` is a model p-value, testing the partial effect of a
(cis-)SNP on gene expression given a set of fixed covariates. The full output
object is a nested list structure. At the first list layer, results are split by
the desired cell type groupings, e.g. "B cells" or "Monocytes". Each of these
groups then is a list of `distQTL` results for each gene under consideration.
Each gene's results consist of a vector of partial $F$ test p-values (presented
as $\log(p)$ results by default), one for each (cis-)SNP tested. As genes can be
differentiaally expressed across different cell types, some genes might be
filtered out for some cell type groups due to low expression; gene level results
generally correspond to the small set of SNPs located near each respective gene.


# Inputs to `distQTL`

## SNP Genotype, Gene Expression, Covariates, and Other Information



## Other Inputs



# Using `distQTL`

## Loading (Part of) the OneK1K Data Set

Part of the OneK1K data set has been included with the `distQTL` package,
specifically expression data from a random selection of 20 genes, donor
genotypes for their various cis-SNPs (defined as within a 200,000 base pair gene
radius), and various supporting information which will be used by the `distQTL`
function.  Loading the package data generates the data objects as exactly
required by `distQTL`.

```{r load_example_OneK1K_Data}
data(package = "distQTL")
```

Let's see the genontype data:

```{r show_genotype}
genotypeDataTable[1:5, 1:5]
```

Let's see the expression data:

```{r show_expression}
expressionDataTable[1:5, 1:5]
```

Finally, let's see the covariate data:

```{r show_covariates}
covariateDataTable[1:5, ]
```

The gene information `geneInfo` and snp information `snpInfo` contain gene and
SNP IDs (respectively), chromosomes, and genome sequence locations (defined as
the start location in the chromosome sequence; for OneK1K, these are with
respect to the **hg19** reference genome).

The `cellTypeGroups` object is a list containing the names of the types of cells
each individual measurement comes from.  The `cellTypeGroups` object included in
the `distQTL` data is a two-part list object, the first containing B cell names:

1. naive B cell,
2. memory B cell,
3. transitional stage B cell,

and the second containing monocyte cell names

1. CD14-positive monocyte,
2. CD14-low, CD16-positive monocyte.

Finally, the object `m` (equal to 200) gives the grid density to evaluate the
empirical distribution objects, given as quantile functions evaluated in
$(0, 1)$; the object `cisRange` (equal to 200,000) gives the range of base pair
values from a gene inside which a SNP is considered a "cis-SNP"; the object
`minCells` (equal to 10) gives the minimum number of cells a donor must have to
be included in the model, evaluated on a cell type group basis; and the object
`minExpr` gives the minimum average positive expression a gene must have, across
donors, to have any models performed on it.

## Fitting distQTL Models

Distribution-based QTL-mapping is performed on this subset of the OneK1K below.

```{r run_distQTL}
t0 = Sys.time()
output = distQTL(genotypeDataTable,
                 expressionDataTable,
                 covariateDataTable,
                 geneInfo,
                 snpInfo,
                 cellTypeGroups,
                 m,
                 cisRange,
                 minCells,
                 minExpr)
runtime = difftime(Sys.time(), t0, units = "sec")
```

This evaluated `{r print_time, echo = FALSE} length(unlist(output))` models in
`{r print_time, echo = FALSE} round(runtime, 1)` seconds, at a model fit rate of
approximately `{r print_time, echo = FALSE} round(runtime/length(unlist(output)), 3)`
seconds per model.

Let's look at results for one gene, in Monocytes:

```{r one_distQTL_gene_example}
DF = data.frame("pval" = round(exp(output[[2]][[3]]), 6))

# The gene:
names(output[[2]])[3]

# Top selection of SNPs:
head(DF[order(DF[ , 1]), , drop = FALSE])
```

